plugins/ptp_operator/metrics/metrics.go

```diff
@@ -239,28 +265,58 @@ func (p *PTPEventManager) ExtractMetrics(msg string) {
 // GenPhc2SysEvent ... generate events form the logs
 func (p *PTPEventManager) GenPhc2SysEvent(iface string, offsetFromMaster float64, clockState ceevent.SyncState) {
 	//TODO: move maxOffsetThreshHold and minOffsetThreshHold to  configmap
-	lastLockState := p.Stats[iface].lastClockState
-	if clockState == ceevent.LOCKED {
-		if offsetFromMaster != 0 { //ptp4l will  return 0 offset
-			if (offsetFromMaster > maxOffsetThreshHold || offsetFromMaster < minOffsetThreshHold) &&
-				lastLockState != ceevent.HOLDOVER {
+	iStats := p.Stats[iface]
+	lastClockState := iStats.lastClockState
+
+	switch clockState {
+	case ceevent.LOCKED:
+		switch lastClockState {
+		case ceevent.FREERUN: //last state was already sent for FreeRUN n, but if its within then send again with new state
+			if offsetFromMaster != 0 && (offsetFromMaster < p.maxOffsetThreshold && offsetFromMaster > p.minOffsetThreshold) { // within range
+				p.publishEvent(clockState, iface) // change to locked
+				iStats.lastClockState = clockState
+				iStats.lastClockStateCount = 1
+			}
+		case ceevent.LOCKED: // last state was in sync , check if its pout of sync
+			if offsetFromMaster == 0 || offsetFromMaster > p.maxOffsetThreshold || offsetFromMaster < p.minOffsetThreshold { // out of sync
 				p.publishEvent(ceevent.HOLDOVER, iface)
-				p.Stats[iface].lastClockState = ceevent.HOLDOVER
-			} else if offsetFromMaster > maxOffsetThreshHold || offsetFromMaster < minOffsetThreshHold &&
-				lastLockState != ceevent.FREERUN {
-				p.Stats[iface].lastClockState = ceevent.FREERUN
-				p.publishEvent(ceevent.FREERUN, iface)
-			} else if lastLockState != clockState {
-				p.publishEvent(clockState, iface)
-				p.Stats[iface].lastClockState = clockState
+				iStats.lastClockState = ceevent.HOLDOVER
+				iStats.lastClockStateCount = 1
 			}
-		} else if clockState != lastLockState {
-			p.Stats[iface].lastClockState = clockState
-			p.publishEvent(clockState, iface)
+		case ceevent.HOLDOVER: // last state was holdover check if it qualifies to be in holdover
+			if offsetFromMaster == 0 || offsetFromMaster > p.maxOffsetThreshold || offsetFromMaster < p.minOffsetThreshold { // still out of sync
+				if iStats.lastClockStateCount < p.holdOverStateThreshold {
+					iStats.lastClockStateCount++
+				} else {
+					p.publishEvent(ceevent.FREERUN, iface)
+					iStats.lastClockState = ceevent.FREERUN
+					iStats.lastClockStateCount = 1
+				}
+			} else {
+				p.publishEvent(clockState, iface) // change to locked
+				iStats.lastClockState = clockState
+				iStats.lastClockStateCount = 1
+			}
+		default: // not yet used states
+			p.publishEvent(clockState, iface) // change to locked
+			iStats.lastClockState = clockState
+			iStats.lastClockStateCount = 1
+			//do nothing already send event
+		}
+	case ceevent.FREERUN:
+		if offsetFromMaster != 0 && (offsetFromMaster < p.maxOffsetThreshold && offsetFromMaster > p.minOffsetThreshold) { // within range
```

Comment:
would look cleaner to have utility functions withinRange(), outOfRange() and lockState()